src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php

```diff
@@ -0,0 +1,219 @@
+    {
+        $schema = match (\count($types = $this->createSubSchemas($node, $definitions, $allowParam))) {
+            1 => $types[0],
+            default => ['anyOf' => $types],
+        };
+
+        if ($node->hasDefaultValue()) {
+            $schema['default'] = $node->getDefaultValue();
+        }
+
+        if ($node instanceof BaseNode) {
+            if ($info = $node->getInfo()) {
+                $schema['description'] = $info;
+            }
+
+            if ($node->isDeprecated()) {
+                $schema['deprecated'] = true;
+                $deprecation = $node->getDeprecation($node->getName(), $node->getPath());
+                $schema['deprecationMessage'] = ($deprecation['package'] || $deprecation['version'] ? "Since {$deprecation['package']} {$deprecation['version']}: " : '').$deprecation['message'];
+            }
+        }
+
+        return $schema;
+    }
+
+    private function createSubSchemas(NodeInterface $node, array &$definitions, bool $allowParam = true): array
+    {
+        [$schema, $validateValue, $allowNull] = match (true) {
+            $node instanceof BooleanNode => [['type' => 'boolean'], is_bool(...), null],
+            $node instanceof IntegerNode => [$this->createNumericSchema($node), is_int(...), null],
+            $node instanceof NumericNode => [$this->createNumericSchema($node), is_float(...), false],
+            $node instanceof StringNode => [['type' => 'string'], is_string(...), ($node->isRequired() && $node->getAllowEmptyValue()) ?: null],
+            $node instanceof EnumNode => [$schema = $this->createEnumSchema($node), static fn ($v) => \in_array($v, $schema['enum']), null],
+            $node instanceof PrototypedArrayNode => [$this->createArraySchema($node, $definitions), static fn () => false, null],
+            $node instanceof ArrayNode => [$this->createObjectSchema($node, $definitions), static fn () => false, null],
+            $node instanceof ScalarNode => [['type' => ['boolean', 'number', 'string']], is_scalar(...), null],
+            default => [[], static fn () => true, null],
+        };
+
+        $allowNull ??= (!$node->isRequired() && ($node->hasDefaultValue() || ($node instanceof VariableNode && $node->getAllowEmptyValue())))
+            || ($node->hasDefaultValue() && null === $node->getDefaultValue());
+
+        $allowedExtraValues = $subSchemas = [];
+        if ($node instanceof BaseNode && $normalizedTypes = $node->getNormalizedTypes()) {
+            $allowedExtraValues = array_column($node->getEquivalentValues(), 0);
+            $allowNull = $allowNull || \in_array(ExprBuilder::TYPE_NULL, $normalizedTypes);
+            if (\in_array(ExprBuilder::TYPE_ANY, $normalizedTypes)) {
+                // This will make IDEs not complain about configurations containing complex beforeNormalization logic
+                $subSchemas[] = new \ArrayObject();
```

Comment:
For an empty object, it's more correct to use `stdClass`. You are using the fact that any object that doesn't expose any property is encoded as an empty object.
```suggestion
                $subSchemas[] = new \stdClass();
```

Reply:
I'm intentionally using `\ArrayObject` (just like [Symfony Serializer does internally](https://github.com/symfony/serializer/blob/7.2/Normalizer/AbstractObjectNormalizer.php#L231)) so that I can perform array operations on it, f.e. `$schema['default'] = $node->getDefaultValue()` and `\count($subSchema, \COUNT_RECURSIVE)`.

If `\stdClass` was used instead, I think the logic would be overall more complex.